internal/dynamic/plugin/config.go

```diff
@@ -10,11 +10,283 @@
+		case FieldTypeBool:
+			f = service.NewBoolField(c.Name)
+		case FieldTypeFloat:
+			f = service.NewFloatField(c.Name)
+		case FieldTypeInt:
+			f = service.NewIntField(c.Name)
+		case FieldTypeString:
+			f = service.NewStringField(c.Name)
+		case FieldTypeUnknown:
+			f = service.NewAnyField(c.Name)
+		default:
+			return nil, fmt.Errorf("unexpected plugin.FieldType: %#v", fieldType)
+		}
+	case FieldKindList:
+		switch fieldType {
+		case FieldTypeBool:
+			// TODO: This should be a BoolListField, but we don't have one yet.
+			f = service.NewAnyListField(c.Name)
+		case FieldTypeFloat:
+			f = service.NewFloatListField(c.Name)
+		case FieldTypeInt:
+			f = service.NewIntListField(c.Name)
+		case FieldTypeString:
+			f = service.NewStringListField(c.Name)
+		case FieldTypeUnknown:
+			f = service.NewAnyListField(c.Name)
+		default:
+			return nil, fmt.Errorf("unexpected plugin.FieldType: %#v", fieldType)
+		}
+	case FieldKindMap:
+		switch fieldType {
+		case FieldTypeBool:
+			// TODO: This should be a BoolMapField, but we don't have one yet.
+			f = service.NewAnyMapField(c.Name)
+		case FieldTypeFloat:
+			f = service.NewFloatMapField(c.Name)
+		case FieldTypeInt:
+			f = service.NewIntMapField(c.Name)
+		case FieldTypeString:
+			f = service.NewStringMapField(c.Name)
+		case FieldTypeUnknown:
+			f = service.NewAnyMapField(c.Name)
+		default:
+			return nil, fmt.Errorf("unexpected plugin.FieldType: %#v", fieldType)
+		}
+	default:
+		return nil, fmt.Errorf("unexpected plugin.FieldKind: %#v", fieldKind)
+	}
+	if c.Default != nil {
+		f = f.Default(*c.Default)
+	}
+	if c.Advanced {
+		f = f.Advanced()
+	}
+	if c.Description != "" {
+		f = f.Description(c.Description)
+	}
+	return f, nil
+}
+
+// Validate checks that the field config is valid.
+func (c *FieldConfig) Validate() error {
+	if c.Name == "" {
+		return errors.New("field name is required")
+	}
+	if c.Type != nil {
+		if err := c.Type.Validate(); err != nil {
+			return err
+		}
+	}
+	if c.Kind != nil {
+		if err := c.Kind.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PluginType describes the type of plugin.
+type PluginType string
+
+// Validate checks that the plugin type is valid.
+func (p PluginType) Validate() error {
+	switch p {
+	case PluginTypeInput, PluginTypeProcessor, PluginTypeOutput:
+		return nil
+	}
+	return fmt.Errorf("invalid plugin type: %q", p)
+}
+
+const (
+	PluginTypeInput     PluginType = "input"
+	PluginTypeProcessor PluginType = "processor"
+	PluginTypeOutput    PluginType = "output"
+)
+
+// Config describes a dynamic plugin over gRPC.
 type Config struct {
+	Name        string `yaml:"name"`
+	Summary     string `yaml:"summary"`
+	Description string `yaml:"description"`
+	// The command to run for the plugin.
+	Cmd    []string      `yaml:"command"`
+	Type   PluginType    `yaml:"type"`
+	Fields []FieldConfig `yaml:"fields"`
+}
+
+// Validate checks that the config is valid.
+func (c *Config) Validate() error {
```

Comment:
It would be cool if we supported declarative validations

Reply:
Can you elaborate what you mean?

Reply:
Oh yeah that's straightforward to do. I was just copying what we supported for templates, but will add a linter field